### ISSUE TITLE
[Sprint] Evaluate and Improve Linear Regression Algorithm

### ISSUE THREAD

**Issue by ankit02327** [ankit02327/stock-price#109]
As part of the Algorithmic & ML Sprint, we need to evaluate, fix, and improve our baseline Linear Regression model (`backend/algorithms/optimised/linear_regression/linear_regression.py`). 

Currently, the implementation serves as a good foundation but has a potential critical bug related to data transformation and lacks regularization techniques which are vital for noisy datasets like stock prices.

**Tasks:**
- [ ] Investigate the `np.log(y)` transformation in the `fit()` and `partial_fit()` methods to ensure it doesn't crash on negative target values (if `y` represents percentage returns).
- [ ] Add support for Regularized Linear Models (Lasso, Ridge, ElasticNet) to prevent overfitting.
- [ ] Evaluate feature importance extraction logic and ensure it scales correctly across batch and SGD training modes.
- [ ] Write robust unit tests covering edge cases (e.g., negative returns, zero variance features).

**Comment by ankit02327** [comment]
1. **Fix Target Transformation:** Remove the `np.log(y)` transformation if the target `y` is a percentage change (which can be negative). If `y` is strictly an absolute price, ensure a validation check `np.any(y <= 0)` is added before applying the log transform.
2. **Introduce Regularization:** Extend the model initialization to accept a `penalty` parameter. If batch learning is used, allow switching between `LinearRegression`, `Ridge`, and `Lasso` from `sklearn.linear_model`. For `SGDRegressor`, configure the `penalty` and `alpha` parameters.
3. **Enhance Feature Importance:** For SGD and Regularized models, ensure that `get_feature_importance()` normalizes coefficients appropriately so they can be interpreted correctly despite scaling.

**Comment by ankit02327** [comment]
```py
# 1. Safe target transformation (if y is price)
if np.any(y <= 0):
    raise ValueError("Target variable contains zero or negative values; cannot apply log transform.")
y_log = np.log(y)

# OR, if y is percentage returns, remove log transform entirely:
# self.model.fit(X_scaled, y) 

# 2. Add Regularization options in __init__
from sklearn.linear_model import Ridge, Lasso, ElasticNet

def __init__(self, use_sgd: bool = False, model_type: str = 'linear', alpha: float = 1.0, **kwargs):
    super().__init__('Linear Regression', **kwargs)
    self.use_sgd = use_sgd
    self.model_type = model_type
    self.alpha = alpha
    
# Inside fit():
if not self.use_sgd:
    if self.model_type == 'ridge':
        self.model = Ridge(alpha=self.alpha)
    elif self.model_type == 'lasso':
        self.model = Lasso(alpha=self.alpha)
    elif self.model_type == 'elasticnet':
        self.model = ElasticNet(alpha=self.alpha)
    else:
        self.model = LinearRegression()
```

**Comment by AKASHPANDEY4642** [comment]
I agree with the proposed solution. I'm working on this.

**Comment by ankit02327** [comment]
Hey you have already claimed 2 issues. I can't assign you any more issues. Please provide the solution for your previously mentioned issues. @AKASHPANDEY4642

**Comment by ankit02327** [comment]
I am assigning this issue to you. But I won't assign you more until I get the PRs related to your previous claims.

**Comment by ankit02327** [comment]
thanks anyways for taking interest @AKASHPANDEY4642 👍
